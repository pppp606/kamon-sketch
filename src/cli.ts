#!/usr/bin/env node

import { hello } from './index';

console.log(hello());