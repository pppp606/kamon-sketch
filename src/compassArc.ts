export type Point = { x: number; y: number }
export type ArcState = 'IDLE' | 'CENTER_SET' | 'RADIUS_SET' | 'DRAWING'

const FULL_CIRCLE_EPS = 0.1
const MIN_RADIUS = 0.001

interface P5DrawingContext {
  push(): void;
  pop(): void;
  noFill(): void;
  stroke(r: number, g?: number, b?: number): void;
  strokeWeight(weight: number): void;
  point(x: number, y: number): void;
  line(x1: number, y1: number, x2: number, y2: number): void;
  circle(x: number, y: number, d: number): void;
  arc(x: number, y: number, w: number, h: number, start: number, stop: number): void;
}

export class CompassArc {
  private centerPoint: Point | null = null
  private radiusPoint: Point | null = null
  private currentPoint: Point | null = null
  private state: ArcState = 'IDLE'
  private lastAngle: number | null = null
  private accumulatedAngle = 0

  getCenterPoint(): Point | null {
    return this.centerPoint
  }

  getRadiusPoint(): Point | null {
    return this.radiusPoint
  }

  getState(): ArcState {
    return this.state
  }

  setCenter(x: number, y: number): void {
    this.centerPoint = { x, y }
    this.radiusPoint = null
    this.currentPoint = null
    this.state = 'CENTER_SET'
    this.resetAngleTracking()
  }

  setRadius(x: number, y: number): void {
    if (!this.centerPoint) {
      throw new Error('Center point must be set before setting radius')
    }
    this.radiusPoint = { x, y }
    this.state = 'RADIUS_SET'
  }

  setRadiusDistance(radius: number): void {
    if (!this.centerPoint) {
      throw new Error('Center point must be set before setting radius')
    }
    // Set radius point at horizontal distance from center (for simplicity)
    this.radiusPoint = { x: this.centerPoint.x + radius, y: this.centerPoint.y }
    this.state = 'RADIUS_SET'
  }

  startDrawing(): void {
    if (this.state !== 'RADIUS_SET') {
      throw new Error('Radius must be set before drawing')
    }
    this.state = 'DRAWING'
    this.resetAngleTracking()
  }

  updateDrawing(x: number, y: number): void {
    if (this.state !== 'DRAWING') {
      throw new Error('Must start drawing before updating')
    }
    this.currentPoint = { x, y }
    this.accumulateAngle()
  }

  getRadius(): number {
    if (!this.centerPoint || !this.radiusPoint) {
      return 0
    }
    const dx = this.radiusPoint.x - this.centerPoint.x
    const dy = this.radiusPoint.y - this.centerPoint.y
    return Math.sqrt(dx * dx + dy * dy)
  }

  private calculateAngle(point: Point): number {
    if (!this.centerPoint) {
      return 0
    }
    return Math.atan2(point.y - this.centerPoint.y, point.x - this.centerPoint.x)
  }

  getStartAngle(): number {
    if (!this.centerPoint || !this.radiusPoint) {
      return 0
    }
    return this.calculateAngle(this.radiusPoint)
  }

  getEndAngle(): number {
    if (!this.centerPoint || !this.currentPoint) {
      return 0
    }
    return this.calculateAngle(this.currentPoint)
  }

  getTotalAngle(): number {
    return this.accumulatedAngle
  }

  isFullCircle(): boolean {
    if (this.state !== 'DRAWING') {
      return false
    }
    
    const totalAngle = this.getTotalAngle()
    
    return Math.abs(totalAngle - 2 * Math.PI) < FULL_CIRCLE_EPS
  }

  draw(p: P5DrawingContext): void {
    if (!this.centerPoint) return

    p.push()
    p.noFill()
    p.stroke(0, 0, 0)
    p.strokeWeight(2)

    if (this.state === 'CENTER_SET') {
      p.point(this.centerPoint.x, this.centerPoint.y)
    } else if (this.state === 'RADIUS_SET' && this.radiusPoint) {
      p.point(this.centerPoint.x, this.centerPoint.y)
      p.line(this.centerPoint.x, this.centerPoint.y, this.radiusPoint.x, this.radiusPoint.y)
      p.circle(this.centerPoint.x, this.centerPoint.y, this.getRadius() * 2)
    } else if (this.state === 'DRAWING' && this.radiusPoint && this.currentPoint) {
      const radius = this.getRadius()
      const startAngle = this.getStartAngle()
      const endAngle = this.getEndAngle()
      
      if (this.isFullCircle()) {
        p.circle(this.centerPoint.x, this.centerPoint.y, radius * 2)
      } else {
        p.arc(this.centerPoint.x, this.centerPoint.y, radius * 2, radius * 2, startAngle, endAngle)
      }
    }

    p.pop()
  }

  reset(): void {
    this.centerPoint = null
    this.radiusPoint = null
    this.currentPoint = null
    this.state = 'IDLE'
    this.resetAngleTracking()
  }

  createCompletedCopy(): CompassArc | null {
    if (this.state !== 'DRAWING' || !this.centerPoint || !this.radiusPoint || !this.currentPoint) {
      return null
    }

    const copy = new CompassArc()
    copy.centerPoint = { ...this.centerPoint }
    copy.radiusPoint = { ...this.radiusPoint }
    copy.currentPoint = { ...this.currentPoint }
    copy.state = 'DRAWING'
    copy.accumulatedAngle = this.accumulatedAngle
    copy.lastAngle = this.lastAngle

    return copy
  }

  private resetAngleTracking(): void {
    this.lastAngle = null
    this.accumulatedAngle = 0
  }

  private accumulateAngle(): void {
    if (!this.centerPoint || !this.currentPoint || !this.radiusPoint) {
      return
    }

    const currentAngle = this.calculateAngle(this.currentPoint)
    const startAngle = this.calculateAngle(this.radiusPoint)
    
    if (this.lastAngle === null) {
      this.lastAngle = startAngle
      this.accumulatedAngle = 0
    }

    let delta = currentAngle - this.lastAngle

    while (delta <= -Math.PI) {
      delta += 2 * Math.PI
    }
    while (delta > Math.PI) {
      delta -= 2 * Math.PI
    }

    this.accumulatedAngle += Math.abs(delta)
    this.lastAngle = currentAngle
  }

  static getFullCircleThreshold(): number {
    return FULL_CIRCLE_EPS
  }

  static getMinRadius(): number {
    return MIN_RADIUS
  }
}