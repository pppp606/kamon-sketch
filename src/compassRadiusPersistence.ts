const DEFAULT_STORAGE_KEY = 'compass.radius'

export class CompassRadiusPersistence {
  private storageKey: string

  constructor(storageKey: string = DEFAULT_STORAGE_KEY) {
    this.storageKey = storageKey
  }

  saveRadius(radius: number): void {
    try {
      if (typeof localStorage !== 'undefined' && localStorage) {
        localStorage.setItem(this.storageKey, radius.toString())
      }
    } catch (error) {
      // Gracefully handle localStorage errors (quota exceeded, disabled, etc.)
      // In production, you might want to log this error
      console.warn('Failed to save compass radius to localStorage:', error)
    }
  }

  loadRadius(): number | null {
    try {
      if (typeof localStorage === 'undefined' || !localStorage) {
        return null
      }

      const storedValue = localStorage.getItem(this.storageKey)
      if (storedValue === null) {
        return null
      }

      // Handle empty or whitespace-only strings
      const trimmedValue = storedValue.trim()
      if (trimmedValue === '') {
        return null
      }

      const parsedValue = parseFloat(trimmedValue)
      
      // Check if parsing resulted in a valid number
      if (isNaN(parsedValue)) {
        return null
      }

      return parsedValue
    } catch (error) {
      // Gracefully handle localStorage errors (security restrictions, etc.)
      console.warn('Failed to load compass radius from localStorage:', error)
      return null
    }
  }

  static getDefaultStorageKey(): string {
    return DEFAULT_STORAGE_KEY
  }
}