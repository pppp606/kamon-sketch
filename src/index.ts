// eslint-disable-next-line @typescript-eslint/no-var-requires
const p5 = require('p5');
import { CompassArc } from './compassArc';
import { Line } from './line';
import { Selection, SelectableElement } from './selection';
import { Fill } from './fill';
import { P5Instance } from './types/p5';

let compassArc: CompassArc;
let completedArcs: CompassArc[] = []; // Store completed arcs
let lines: Line[] = [];
let currentLine: Line | null = null;
let drawingMode: 'compass' | 'line' | 'fill' = 'line'; // Default to line mode for MVP
let selection: Selection;
let fill: Fill;

export function hello(): string {
  return 'Hello World';
}

export function setDrawingMode(mode: 'compass' | 'line' | 'fill'): void {
  drawingMode = mode;
  // Reset current drawings when switching modes
  if (compassArc) {
    compassArc.reset();
  }
  currentLine = null;
}

export function getDrawingMode(): 'compass' | 'line' | 'fill' {
  return drawingMode;
}

export function getLines(): Line[] {
  return lines;
}

export function getCurrentLine(): Line | null {
  return currentLine;
}

export function setup(p: P5Instance): void {
  p.createCanvas(800, 800);
  p.background(220);
  compassArc = new CompassArc();
  completedArcs = [];
  lines = [];
  currentLine = null;
  selection = new Selection();
  fill = new Fill();
}

export function draw(p: P5Instance): void {
  p.clear();
  p.background(220);
  
  // Draw all completed lines
  for (const line of lines) {
    line.draw(p);
  }
  
  // Draw all completed arcs
  for (const arc of completedArcs) {
    arc.draw(p);
  }
  
  // Draw current line being created
  if (currentLine) {
    currentLine.draw(p);
  }
  
  // Draw the compass arc (if in compass mode)
  if (drawingMode === 'compass' && compassArc) {
    compassArc.draw(p);
  }
  
  // Draw selection highlight
  if (selection) {
    selection.drawHighlight(p as Parameters<typeof selection.drawHighlight>[0]);
  }
}

export function mousePressed(p: P5Instance): void {
  // Handle fill mode first
  if (drawingMode === 'fill' && fill) {
    fill.handleClick(p, p.mouseX, p.mouseY);
    return;
  }
  
  // Handle compass arc drawing first (it has precedence when in progress)
  if (drawingMode === 'compass' && compassArc) {
    const state = compassArc.getState();
    
    if (state === 'IDLE') {
      selection.setSelectedElement(null); // Clear selection when starting new compass arc
      compassArc.setCenter(p.mouseX, p.mouseY);
      return;
    } else if (state === 'CENTER_SET') {
      selection.setSelectedElement(null); // Clear selection when setting radius
      compassArc.setRadius(p.mouseX, p.mouseY);
      return;
    }
    // If state is DRAWING, continue to selection logic
  }
  
  // Selection logic (only when not in middle of compass arc creation)
  const selectableElements: SelectableElement[] = [
    ...lines.map(line => ({ type: 'line' as const, element: line })),
    ...(compassArc && (compassArc.getState() === 'DRAWING' || compassArc.getState() === 'RADIUS_SET') ? [{ type: 'arc' as const, element: compassArc }] : [])
  ];
  
  const clickPoint = { x: p.mouseX, y: p.mouseY };
  const closestElement = selection.findClosestElement(clickPoint, selectableElements);
  
  // Selection threshold - only select if click is close enough
  const SELECTION_THRESHOLD = 20;
  
  if (closestElement) {
    let distance: number;
    if (closestElement.type === 'line') {
      distance = selection.calculateDistanceToLine(clickPoint, closestElement.element);
    } else {
      distance = selection.calculateDistanceToArc(clickPoint, closestElement.element);
    }
    
    if (distance <= SELECTION_THRESHOLD) {
      selection.setSelectedElement(closestElement);
      return; // Don't proceed with drawing when selecting
    }
  }
  
  // Clear selection if not clicking on an element
  selection.setSelectedElement(null);
  
  // Proceed with line drawing logic
  if (drawingMode === 'line') {
    if (!currentLine) {
      // Start a new line
      currentLine = new Line();
      currentLine.setFirstPoint(p.mouseX, p.mouseY);
    } else if (currentLine.getState() === 'FIRST_POINT') {
      // Set the second point and complete the line
      currentLine.setSecondPoint(p.mouseX, p.mouseY);
      lines.push(currentLine);
      // Start a new line immediately
      currentLine = new Line();
      currentLine.setFirstPoint(p.mouseX, p.mouseY);
    }
  }
}

export function mouseDragged(p: P5Instance): void {
  if (drawingMode === 'compass' && compassArc) {
    if (compassArc.getState() === 'RADIUS_SET') {
      // Start drawing when dragging begins from RADIUS_SET state
      compassArc.startDrawing();
      compassArc.updateDrawing(p.mouseX, p.mouseY);
    } else if (compassArc.getState() === 'DRAWING') {
      compassArc.updateDrawing(p.mouseX, p.mouseY);
    }
  }
  // Line mode doesn't use drag for drawing (click-click interaction)
}

export function mouseReleased(): void {
  if (drawingMode === 'compass' && compassArc) {
    if (compassArc.getState() === 'DRAWING') {
      // Create a complete copy of the current arc and add to completed arcs
      const completedArc = new CompassArc();
      completedArc.copyFrom(compassArc);
      completedArcs.push(completedArc);
      compassArc.reset();
    }
  }
  // Line mode doesn't use mouse release events
}

export function getCompassArc() {
  return compassArc;
}

export function getSelection() {
  return selection;
}

export function getFill() {
  return fill;
}

export function setFillColor(color: { r: number; g: number; b: number }) {
  if (fill) {
    fill.setFillColor(color);
  }
}

export function getFillColor() {
  return fill ? fill.getFillColor() : { r: 0, g: 0, b: 0 };
}

export function doubleClicked(p: P5Instance): void {
  const selectedElement = selection.getSelectedElement()
  
  if (!selectedElement) {
    return
  }

  const clickPoint = { x: p.mouseX, y: p.mouseY }
  const closestPoint = selection.getClosestPointOnElement(clickPoint, selectedElement)
  
  if (!closestPoint) {
    return
  }

  // Start new drawing from the closest point on the selected element
  if (drawingMode === 'line') {
    // Clear any current line and start new one from the closest point
    currentLine = new Line()
    currentLine.setFirstPoint(closestPoint.x, closestPoint.y)
  } else if (drawingMode === 'compass') {
    // For compass mode, start new arc with center at the closest point
    compassArc.reset()
    compassArc.setCenter(closestPoint.x, closestPoint.y)
  }
  
  // Clear selection after starting new drawing
  selection.setSelectedElement(null)
}

export function startDrawingFromSelectedElement(): boolean {
  const selectedElement = selection.getSelectedElement()
  
  if (!selectedElement) {
    return false
  }

  // For line mode, start a new line from the closest point on the selected element
  if (drawingMode === 'line') {
    // Clear any current line
    currentLine = null
    return true
  }
  
  // For compass mode, we would implement arc-to-arc continuation here
  return false
}

export function createSketch(): void {
  new p5((p: P5Instance) => {
    p.setup = () => setup(p);
    p.draw = () => draw(p);
    p.mousePressed = () => mousePressed(p);
    p.mouseDragged = () => mouseDragged(p);
    p.mouseReleased = () => mouseReleased();
    p.doubleClicked = () => doubleClicked(p);
  });
}

// Auto-initialize if running in browser
if (typeof window !== 'undefined') {
  createSketch();
}