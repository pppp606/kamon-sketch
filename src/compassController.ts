import { CompassArc, Point } from './compassArc'
import { CompassRadiusState } from './compassRadiusState'
import { CompassRadiusPersistence } from './compassRadiusPersistence'

interface P5DrawingContext {
  push(): void;
  pop(): void;
  noFill(): void;
  stroke(r: number, g?: number, b?: number): void;
  strokeWeight(weight: number): void;
  point(x: number, y: number): void;
  line(x1: number, y1: number, x2: number, y2: number): void;
  circle(x: number, y: number, d: number): void;
  arc(x: number, y: number, w: number, h: number, start: number, stop: number): void;
}

interface MouseEvent {
  shiftKey: boolean;
}

export type CompassControllerState = 'IDLE' | 'CENTER_SET' | 'DRAWING' | 'SETTING_RADIUS'

export class CompassController {
  private compassArc: CompassArc
  private radiusState: CompassRadiusState
  private persistence: CompassRadiusPersistence
  private currentMouseX: number = 0
  private currentMouseY: number = 0

  constructor() {
    this.compassArc = new CompassArc()
    this.persistence = new CompassRadiusPersistence()
    
    // Load radius from localStorage or use default
    const savedRadius = this.persistence.loadRadius()
    const initialRadius = savedRadius !== null ? savedRadius : 10
    this.radiusState = new CompassRadiusState(initialRadius)
  }

  getCurrentRadius(): number {
    return this.radiusState.getCurrentRadius()
  }

  getLastRadius(): number {
    return this.radiusState.getLastRadius()
  }

  setCurrentRadius(radius: number): void {
    this.radiusState.setCurrentRadius(radius)
    this.persistence.saveRadius(radius)
  }

  getCenterPoint(): Point | null {
    return this.compassArc.getCenterPoint()
  }

  getCompassCenter(): Point | null {
    return this.radiusState.getCompassCenter()
  }

  getState(): CompassControllerState {
    const radiusStateType = this.radiusState.getState()
    if (radiusStateType === 'SETTING_RADIUS') {
      return 'SETTING_RADIUS'
    }

    const arcState = this.compassArc.getState()
    switch (arcState) {
      case 'IDLE': return 'IDLE'
      case 'CENTER_SET': return 'CENTER_SET'
      case 'RADIUS_SET': return 'CENTER_SET' // Treat as center set since we use persistent radius
      case 'DRAWING': return 'DRAWING'
      default: return 'IDLE'
    }
  }

  handleClick(x: number, y: number, event: MouseEvent): void {
    this.currentMouseX = x
    this.currentMouseY = y

    if (event.shiftKey) {
      this.handleShiftClick(x, y)
    } else {
      this.handleNormalClick(x, y)
    }
  }

  handleMouseDrag(x: number, y: number): void {
    this.currentMouseX = x
    this.currentMouseY = y

    const radiusState = this.radiusState.getState()
    if (radiusState === 'SETTING_RADIUS') {
      // Update radius based on distance from compass center
      this.radiusState.updateShiftRadiusFromPoint({ x, y })
    } else {
      // Handle normal compass arc drawing
      const arcState = this.compassArc.getState()
      if (arcState === 'DRAWING') {
        this.compassArc.updateDrawing(x, y)
      }
    }
  }

  handleMouseRelease(x: number, y: number): void {
    this.currentMouseX = x
    this.currentMouseY = y

    const radiusState = this.radiusState.getState()
    if (radiusState === 'SETTING_RADIUS') {
      // Finalize radius setting
      this.radiusState.finalizeShiftOperation()
      this.persistence.saveRadius(this.radiusState.getCurrentRadius())
    }

    // Check for full circle completion in normal drawing mode
    const arcState = this.compassArc.getState()
    if (arcState === 'DRAWING' && this.compassArc.isFullCircle()) {
      this.reset()
    }
  }

  handleKeyPress(key: string): void {
    if (key === 'Escape') {
      this.cancelCurrentOperation()
    }
  }

  handleRightClick(): void {
    this.cancelCurrentOperation()
  }

  draw(p: P5DrawingContext): void {
    const radiusState = this.radiusState.getState()
    
    if (radiusState === 'SETTING_RADIUS') {
      this.drawRadiusSetting(p)
    } else {
      this.drawNormalCompass(p)
    }
  }

  reset(): void {
    this.compassArc.reset()
    // Don't reset radius state - preserve current radius
  }

  getCompassArc(): CompassArc {
    return this.compassArc
  }

  private handleShiftClick(x: number, y: number): void {
    const radiusState = this.radiusState.getState()
    
    if (radiusState === 'IDLE') {
      // Start radius setting operation
      this.radiusState.startShiftOperation({ x, y })
    } else {
      // Cancel ongoing operation
      this.cancelCurrentOperation()
    }
  }

  private handleNormalClick(x: number, y: number): void {
    const arcState = this.compassArc.getState()
    
    switch (arcState) {
      case 'IDLE':
        // Set center point
        this.compassArc.setCenter(x, y)
        break
        
      case 'CENTER_SET':
        // Use current radius to calculate radius point, then start drawing
        this.setRadiusFromCurrent(x, y)
        this.compassArc.startDrawing()
        break
        
      case 'DRAWING':
        // Complete current drawing and start new one
        this.reset()
        this.compassArc.setCenter(x, y)
        break
        
      default:
        break
    }
  }

  private setRadiusFromCurrent(clickX: number, clickY: number): void {
    const center = this.compassArc.getCenterPoint()
    if (!center) return

    const currentRadius = this.radiusState.getCurrentRadius()
    
    // Calculate direction from center to click point
    const dx = clickX - center.x
    const dy = clickY - center.y
    const distance = Math.sqrt(dx * dx + dy * dy)
    
    if (distance === 0) {
      // If clicked on center, use default direction (right)
      this.compassArc.setRadius(center.x + currentRadius, center.y)
    } else {
      // Scale the direction to match current radius
      const scale = currentRadius / distance
      const radiusX = center.x + dx * scale
      const radiusY = center.y + dy * scale
      this.compassArc.setRadius(radiusX, radiusY)
    }
  }

  private drawRadiusSetting(p: P5DrawingContext): void {
    const compassCenter = this.radiusState.getCompassCenter()
    if (!compassCenter) return

    p.push()
    p.noFill()
    p.stroke(0, 150, 255) // Blue color for radius setting mode
    p.strokeWeight(2)

    // Draw center point
    p.point(compassCenter.x, compassCenter.y)
    
    // Draw radius line to current mouse position
    p.line(compassCenter.x, compassCenter.y, this.currentMouseX, this.currentMouseY)
    
    // Draw preview circle with current radius
    const currentRadius = this.radiusState.getCurrentRadius()
    p.circle(compassCenter.x, compassCenter.y, currentRadius * 2)
    
    p.pop()
  }

  private drawNormalCompass(p: P5DrawingContext): void {
    // Use the existing CompassArc drawing logic
    this.compassArc.draw(p)
  }

  private cancelCurrentOperation(): void {
    const radiusState = this.radiusState.getState()
    
    if (radiusState === 'SETTING_RADIUS') {
      this.radiusState.cancelShiftOperation()
    } else {
      // Cancel normal compass operation
      this.reset()
    }
  }
}