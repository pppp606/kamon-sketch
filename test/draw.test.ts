import 'jest-canvas-mock';
import { setup, draw, createSketch } from '../src/index';

interface P5Instance {
  createCanvas: jest.Mock;
  background: jest.Mock;
  clear: jest.Mock;
  stroke: jest.Mock;
  strokeWeight: jest.Mock;
  line: jest.Mock;
  fill: jest.Mock;
  ellipse: jest.Mock;
}

describe('p5.js Drawing Tests', () => {
  let p: P5Instance;
  let canvas: HTMLCanvasElement;
  let ctx: CanvasRenderingContext2D;

  beforeEach(() => {
    // Create a mock p5 instance
    canvas = document.createElement('canvas');
    canvas.width = 400;
    canvas.height = 400;
    ctx = canvas.getContext('2d') as CanvasRenderingContext2D;

    // Mock p5 methods
    p = {
      createCanvas: jest.fn().mockReturnValue(canvas),
      background: jest.fn(),
      clear: jest.fn(),
      stroke: jest.fn(),
      strokeWeight: jest.fn(),
      line: jest.fn(),
      fill: jest.fn(),
      ellipse: jest.fn(),
    } as any;
  });

  test('should create canvas with correct dimensions', () => {
    setup(p);
    expect(p.createCanvas).toHaveBeenCalledWith(400, 400);
    expect(p.background).toHaveBeenCalledWith(220);
  });

  test('should draw line with correct parameters', () => {
    draw(p);
    
    expect(p.clear).toHaveBeenCalled();
    expect(p.background).toHaveBeenCalledWith(220);
    
    // Verify line drawing
    expect(p.stroke).toHaveBeenCalledWith(255, 0, 0);
    expect(p.strokeWeight).toHaveBeenCalledWith(2);
    expect(p.line).toHaveBeenCalledWith(50, 50, 350, 350);
  });

  test('should draw circle with correct parameters', () => {
    draw(p);
    
    // Verify circle drawing
    expect(p.fill).toHaveBeenCalledWith(0, 255, 0);
    expect(p.stroke).toHaveBeenCalledWith(0);
    expect(p.strokeWeight).toHaveBeenCalledWith(1);
    expect(p.ellipse).toHaveBeenCalledWith(200, 200, 100, 100);
  });

  test('should clear canvas before drawing', () => {
    draw(p);
    
    expect(p.clear).toHaveBeenCalled();
    expect(p.background).toHaveBeenCalledWith(220);
  });

  test('canvas drawing operations snapshot', () => {
    // Create a real canvas context for snapshot testing
    const realCanvas = document.createElement('canvas');
    realCanvas.width = 400;
    realCanvas.height = 400;
    const realCtx = realCanvas.getContext('2d') as CanvasRenderingContext2D;

    // Simulate drawing operations manually for snapshot
    realCtx.clearRect(0, 0, 400, 400);
    realCtx.fillStyle = '#dcdcdc'; // background(220) in p5.js
    realCtx.fillRect(0, 0, 400, 400);
    
    // Draw line (red stroke)
    realCtx.strokeStyle = 'rgb(255, 0, 0)';
    realCtx.lineWidth = 2;
    realCtx.beginPath();
    realCtx.moveTo(50, 50);
    realCtx.lineTo(350, 350);
    realCtx.stroke();
    
    // Draw circle (green fill, black stroke)
    realCtx.fillStyle = 'rgb(0, 255, 0)';
    realCtx.strokeStyle = 'rgb(0, 0, 0)';
    realCtx.lineWidth = 1;
    realCtx.beginPath();
    realCtx.arc(200, 200, 50, 0, 2 * Math.PI);
    realCtx.fill();
    realCtx.stroke();

    // Take snapshot of canvas operations
    const events = (realCtx as any).__getEvents();
    expect(events).toMatchSnapshot();
  });

  test('should create p5.js sketch instance', () => {
    // Test that the function exists and is callable
    expect(typeof createSketch).toBe('function');
    
    // Should not throw errors when called
    expect(() => createSketch()).not.toThrow();
  });
});