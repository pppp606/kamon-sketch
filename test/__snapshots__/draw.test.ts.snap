// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p5.js Drawing Tests canvas drawing operations snapshot 1`] = `
[
  {
    "props": {
      "height": 400,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "value": "#dcdcdc",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  {
    "props": {
      "height": 400,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  {
    "props": {
      "value": "#ff0000",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  {
    "props": {
      "value": 2,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  {
    "props": {
      "x": 50,
      "y": 50,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  {
    "props": {
      "x": 350,
      "y": 350,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        {
          "props": {
            "x": 50,
            "y": 50,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        {
          "props": {
            "x": 350,
            "y": 350,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  {
    "props": {
      "value": "#00ff00",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  {
    "props": {
      "value": "#000000",
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  {
    "props": {
      "value": 1,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  {
    "props": {
      "anticlockwise": false,
      "endAngle": 6.283185307179586,
      "radius": 50,
      "startAngle": 0,
      "x": 200,
      "y": 200,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  {
    "props": {
      "fillRule": "nonzero",
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        {
          "props": {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 50,
            "startAngle": 0,
            "x": 200,
            "y": 200,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  {
    "props": {
      "path": [
        {
          "props": {},
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        {
          "props": {
            "anticlockwise": false,
            "endAngle": 6.283185307179586,
            "radius": 50,
            "startAngle": 0,
            "x": 200,
            "y": 200,
          },
          "transform": [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
